src/theory/ff/model.cpp

```diff
@@ -0,0 +1,303 @@
+          * vars.size())
+{
+}
+
+RoundRobinEnumerator::~RoundRobinEnumerator() {}
+
+std::optional<CoCoA::RingElem> RoundRobinEnumerator::next()
+{
+  std::optional<CoCoA::RingElem> ret{};
+  if (d_idx != d_maxIdx)
+  {
+    size_t whichVar = d_idx % d_vars.size();
+    CoCoA::BigInt whichVal = d_idx / d_vars.size();
+    CoCoA::RingElem val = d_ring->myZero();
+    val += whichVal;
+    ret = d_vars[whichVar] - val;
+    ++d_idx;
+  }
+  return ret;
+}
+
+std::string RoundRobinEnumerator::name() { return "round-robin"; }
+
+bool isUnsat(const CoCoA::ideal& ideal)
+{
+  const auto& gens = CoCoA::GBasis(ideal);
+  return !(gens.size() > 1 || CoCoA::deg(gens[0]) > 0);
+}
+
+template <typename T>
+std::string ostring(const T& t)
+{
+  std::ostringstream o;
+  o << t;
+  return o.str();
+}
+
+std::pair<size_t, CoCoA::RingElem> extractAssignment(
+    const CoCoA::RingElem& elem)
+{
+  Assert(CoCoA::deg(elem) == 1);
+  Assert(CoCoA::NumTerms(elem) <= 2);
+  CoCoA::RingElem m = CoCoA::monic(elem);
+  int varNumber = CoCoA::UnivariateIndetIndex(elem);
+  Assert(varNumber >= 0);
+  return {varNumber, -CoCoA::ConstantCoeff(m)};
+}
+
+std::unordered_set<std::string> assignedVars(const CoCoA::ideal& ideal)
+{
+  std::unordered_set<std::string> ret{};
+  for (const auto& g : CoCoA::GBasis(ideal))
+  {
+    if (CoCoA::deg(g) == 1)
+    {
+      int varNumber = CoCoA::UnivariateIndetIndex(g);
+      if (varNumber >= 0)
+      {
+        ret.insert(ostring(CoCoA::indet(ideal->myRing(), varNumber)));
+      }
+    }
+  }
+  return ret;
+}
+
+bool allVarsAssigned(const CoCoA::ideal& ideal)
+{
+  return assignedVars(ideal).size()
+         == (size_t)CoCoA::NumIndets(ideal->myRing());
+}
+
+std::unique_ptr<AssignmentEnumerator> brancher(const CoCoA::ideal& ideal)
+{
+  CoCoA::ring polyRing = ideal->myRing();
+  Assert(!isUnsat(ideal));
+  // first, we look for super-linear univariate polynomials.
+  const auto& gens = CoCoA::GBasis(ideal);
+  for (const auto& p : gens)
+  {
+    int varNumber = CoCoA::UnivariateIndetIndex(p);
+    if (varNumber >= 0 && CoCoA::deg(p) > 1)
+    {
+      return factorEnumerator(p);
+    }
+  }
+  // now, we check the dimension
+  if (CoCoA::IsZeroDim(ideal))
+  {
+    // If zero-dimensional, we compute a minimal polynomial in some unset
+    // variable.
+    std::unordered_set<std::string> alreadySet = assignedVars(ideal);
+    for (const auto& var : CoCoA::indets(polyRing))
+    {
+      std::string varName = ostring(var);
+      if (!alreadySet.count(ostring(var)))
+      {
+        CoCoA::RingElem minPoly = CoCoA::MinPolyQuot(var, ideal, var);
+        return factorEnumerator(minPoly);
+      }
+    }
+    Unreachable()
+        << "There should be no unset variables in zero-dimensional ideal";
+  }
+  else
+  {
+    // If positive dimensional, we make a list of unset variables and
+    // round-robin guess.
+    //
+    // TODO(aozdemir): guess free variables first!
```

Comment:
We require wishues for TODOs like this in code, can you open and reference the number?